test/util/net/netlink.go

```diff
@@ -0,0 +1,364 @@
+package net
+
+import (
+	"fmt"
+	"net"
+	"runtime"
+
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+
+	"github.com/k8snetworkplumbingwg/sriov-cni/test/util/ethtool"
+)
+
+const (
+	// LinkStateAuto - Link state auto
+	LinkStateAuto = 0
+	// LinkStateEnable - Link state enable
+	LinkStateEnable = 1
+	// LinkStateDisable - Link state disable
+	LinkStateDisable = 2
+)
+
+// SetVfLinkState set VF link state
+// state 0 - auto (linkStateAuto), 1 - enable (linkStateEnable), 2 - disable (linkStateDisable) - defined as constants
+func SetVfLinkState(pfName, containerNs string, vf int, state uint32) error {
+	// Get container network namespace from path
+	ns, err := netns.GetFromPath(containerNs)
+	if nil != err {
+		return err
+	}
+	defer ns.Close()
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	hostNetNs, err := netns.Get() // host network namespace
+	if err != nil {
+		return err
+	}
+
+	err = netns.Set(ns) // switch netns to container namespace
+	if err != nil {
+		return err
+	}
+
+	link, err := netlink.LinkByName(pfName)
+	if nil != err {
+		return err
+	}
+
+	err = netlink.LinkSetVfState(link, vf, state)
+	if nil != err {
+		return err
+	}
+
+	// Return to the host network namespace
+	err = netns.Set(hostNetNs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SetAllVfOnLinkState sets all VFs of given PF to selected state
+func SetAllVfOnLinkState(pfName, containerNs string, state uint32) error {
+	links, err := GetVfsLinksInfoList(pfName, containerNs)
+
+	if err != nil {
+		return err
+	}
+
+	for vfIdx := 0; vfIdx < len(links); vfIdx++ {
+		err = SetVfLinkState(pfName, containerNs, vfIdx, state)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// SetVfsMAC set MACs on VFs to defined state starting from 4a:ea:39:09:4e:XX
+// NOTE: Loop will generate only last two positions
+func SetVfsMAC(pfName, containerNs string) error {
+	// Get container network namespace from path
+	ns, err := netns.GetFromPath(containerNs)
+	if nil != err {
+		return err
+	}
+	defer ns.Close()
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	hostNetNs, err := netns.Get() // host network namespace
+	if err != nil {
+		return err
+	}
+	err = netns.Set(ns) // switch netns to container namespace
+	if err != nil {
+		return err
+	}
+
+	link, err := netlink.LinkByName(pfName)
+	if nil != err {
+		return err
+	}
+
+	var baseMac string = "4a:ea:39:09:4e:"
+	for index := 0; index < len(link.Attrs().Vfs); index++ {
+		var mac string
+		if index < 17 {
+			mac = baseMac + fmt.Sprintf("0%x", index)
+		} else {
+			mac = baseMac + fmt.Sprintf("%x", index)
+		}
+
+		hwMac, err := net.ParseMAC(mac)
+		if err != nil {
+			return err
+		}
+
+		err = netlink.LinkSetVfHardwareAddr(link, index, hwMac)
+		if nil != err {
+			return err
+		}
+	}
+
+	// Return to the host network namespace
+	err = netns.Set(hostNetNs)
```

Comment:
Isn't it safer to do it in a `defer ...` statement?

Reply:
Should we always return to host namespace if another error occurs earlier?

Reply:
Agreed.

Reply:
removed, and added a check in case an error occurred when return to host network namespace.